Create and edit nodes in temporary ast nodes with execution context disabled

vitvakatu opened the pull request:
### Pull Request Description

Closes #6417

Creating and editing nodes with a component browser does not edit the node in place but creates a temporary AST node with the execution context switch. When the node is committed, the original AST is replaced by the new one, and the temporary node is removed.

This solves several issues:
1. Previewing suggestions or editing nodes should no longer produce unwanted side effects, as the edited node expression is always evaluated with the output context being disabled. 
2. Errors no longer propagate down the graph when you edit the node. The edited node has no outputs connected. Inputs behave normally, though.
3. Undo-redo and editing abort is much simpler. To abort editing, remove the temporary node.

https://github.com/enso-org/enso/assets/6566674/8d38e6bc-a4de-4e69-8d48-52d9b45b6554

### Important Notes

The code is a bit awkward in a few places. This is the consequence of previous architecture.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

vitvakatu committed:
Create and edit nodes in temporary ast nodes with execution context disabled

farmaazon commented on app/gui/src/controller/searcher.rs:
```diff
@@ -514,12 +512,6 @@ impl Searcher {
     /// If `suggestion` is specified, the preview will contains code after applying it.
     /// Otherwise it will be just the current searcher input.
     pub fn preview(&self, suggestion: Option<&component::Suggestion>) -> FallibleResult {
-        let transaction_name = "Previewing Component Browser suggestion.";
-        let _skip = self
-            .graph
-            .undo_redo_repository()
-            .open_ignored_transaction_or_ignore_current(transaction_name);
-
```
I don't understand why we do not need this anymore. Updating even temporary node updates the module, what in turn will create a transaction if we won't ignore it somewhere.

vitvakatu replied:
Now, we open the transaction once in `ComponentBrowserSearcher::setup_searcher`, and close it when the searcher presenter is dropped. When the user wants to discard the edit, we ignore the transaction (see `ComponentBrowserSearcher::abort_editing`). This way, we have a single transaction (I call it _Transaction_ in this comment) for the whole lifetime of the CB, and as far as I understand, that is what we need from the user perspective.

We can't use `open_ignored_transaction_or_ignore_current` here and in other methods of the searcher controller because it will ignore that _Transaction_ we opened in the presenter. Instead, we could use `let _skip = self.graph.get_or_open_transaction(transaction_name)`, and the `transaction_name` will be appended to the _Transaction_ log. I find it unnecessary, though.

If we remove the _Transaction_ from the presenter, we will have a bunch of individual transactions created by different searcher controller methods, like this one. Undoing node edit will then require undoing multiple stack frames, which is definitely not what we want.

farmaazon replied:
The problem is when something else is changed when CB is opened. For example: connection is added (but this is a strange action and may be forbidden) or opening visualization (this in turn is quite of UX improvement). This action would be then "merged" with CB action and discarded if aborting node adding/editing.

vitvakatu replied:
Shouldn't these actions create their own transactions? The order might be a little off, though. Like when you toggle visualization with the CB being opened, you will first discard the edit and only then revert the visualization change. 

I don't see how we can mitigate this. If we split CB actions to separate transactions it's not clear to me how merge them into a single undo frame.

vitvakatu replied:
After a call we decided to try restore the old behavior of transactions but utilizing existing infrastructure for removing temporary code from undo stack. I'll check how hard it is.

farmaazon commented on app/gui/src/presenter/graph/state.rs:
```diff
@@ -117,16 +118,22 @@ impl Nodes {
+        let old_ast_to_remove = if let Some(old_view) = displayed.view_id {
+            self.ast_node_by_view_id.remove(&old_view)
         } else {
-            self.nodes_without_view.remove_item(&ast_id);
```
If we assign a node explicitly, and it had no view, then it should still be removed from "nodes_without_view" list, shouldn't it?

vitvakatu replied:
I guess you're right.

My initial thought process was the following:
- We had this line in the previous version of the function because of `get_mut_or_create_static`, which adds a new node to the `nodes_without_view` list.
- We can safely skip this step because we don't use `get_mut_or_create_static` anymore.

Now I see that there are possible situations when we create the node by other means and only then connect it to the view using this method. I didn't find any such code path in our current implementation, but I guess this can change in the future. After all, removing the item is a safe action here. I added it back in the commit 939549b0dd7d3a671402dd8d0ea02dbb0e94a930

farmaazon reviewed (commented):
There are some parts I don't know why they are working.

vitvakatu committed:
Remove node from nodes_without_view list for safety

vitvakatu commented:
For future reference, here is what happened after the discussion in https://github.com/enso-org/enso/pull/7779#discussion_r1321184973

You can see the changes I made to the branch here: https://github.com/enso-org/enso/compare/develop...vitvakatu:enso:wip/vitvakatu/preview-context-switch-attempt-to-fix-undo-redo

I did as we discussed:
1. Returned the transactions removed from the searcher controller
2. Stored NodeEditStatus::Created for the temporary node, so it does not appear in snapshots
3. Instead of a huge transaction for the whole lifetime of the CB, I use only a transaction on commit.

I tested the changes by the following actions:
1. Have one node without visualization on the scene
2. Open a CB for creating another node
3. Open visualization on the first node
4. Accept new node
5. Press Undo to discard node creation
6. Press Undo a second time to discard the visualization toggle.
Expected:
The first Undo removes a new node, the second Undo closes visualization.

It all happened in the first five commits (up until c00a5bb964d57c0879fae621c9daecdfe7bb88f9)

After doing so, I noticed that the fix is not working: visualization is still closing after the first Undo, as before. Moreover, we were creating an additional undo frame (seemingly empty), so a third Undo was needed to reach the end of the stack.

I debugged and fixed the second issue: showing a preview for the node also creates an undo frame (which is a bit weird, but well). So I simply ignored this additional frame by moving the transaction from create_input_node to init_node. 
Then, I noticed that accepting a node does not create a stack frame (wtf?). Turns out, we can’t use `open_ignored_transaction_or_ignore_current` anywhere while committing the node! That’s because everything happens inside a large transaction created for the `accept_selected_entry` shortcut, and ignoring it will ignore the whole commit.

After addressing these two points (it happened in the last commit), I finally got to the point where everything worked great… until you try to repeat it one more time.

See the attached video:


https://github.com/enso-org/enso/assets/6566674/979dbe55-f895-4e1a-a73e-b6bbc901cf47



1. I toggle visualization, notice that it creates undo frames each time, which is correct.
2. I did the test outline above once. Everything works exactly as desired. No redundant frames; everything’s great.
3. I try to repeat my actions again – and visualization does not close.
4. Toggling visualization no longer creates frames.

---

As a result, we decided to keep the current implementation as in PR.

farmaazon commented:
### QA Report: :red_circle: 

1. When no entry in CB is selected, the preview should always show the result of the current input. But when I deselect entry in CB (clicking on input, focusing it), the preview is not updated. Also, having good input, but error in preview due to this bug while editing an existing node, when I accept input with cmd+enter the error stays on the node.

https://github.com/enso-org/enso/assets/3919101/841918fc-0870-47a6-a199-a29f602a87f0

2. When I create a new node with source (e.g. by dragging out an edge), there is no edge from source node to CB input.

mwu-tow committed:
wip

vitvakatu committed:
Update preview even when no suggestion is selected

mwu-tow committed:
wip

mwu-tow committed:
wip

mwu-tow committed:
Merge remote-tracking branch 'origin/develop' into wip/mwu/new-ide-build

vitvakatu committed:
Fix the issue with missing input edge for new nodes

vitvakatu committed:
Merge branch 'develop' into wip/vitvakatu/preview-context-switch

vitvakatu commented:
> When no entry in CB is selected, the preview should always show the result of the current input. But when I deselect entry in CB (clicking on input, focusing it), the preview is not updated. Also, having good input, but error in preview due to this bug while editing an existing node, when I accept input with cmd+enter the error stays on the node.

This issue is present on develop as well (I didn't notice it immediately). It happens because updating preview is tied to selecting suggestion from the list. I fixed it in [2c46dab](https://github.com/enso-org/enso/pull/7779/commits/2c46dab7030f91bbce415bb11a4081eefad27119) by triggering the preview update when the input is changed with no suggestions selected in the list.

> When I create a new node with source (e.g. by dragging out an edge), there is no edge from source node to CB input.

Turns out it works on develop purely by accident. I'm not 100% sure, but it seems on develop we simply don't have any input port, and thus use a default port offset for the edge target position. On my branch, we now generate a context switch expression that has a port, but the port is not visible, and its transform is not updated. I fixed it in [21f63d9](https://github.com/enso-org/enso/pull/7779/commits/21f63d93b879fd6fd93f2107845fa8b0c4591d84) by choosing a default offset when the returned position is zero (which probably won't ever happen in other cases).

mwu-tow committed:
wip

farmaazon commented:
### QA Report: :red_square: 

When I write bad input and then cancels the edit with esc, the error stays on the node, apparently forever.

https://github.com/enso-org/enso/assets/3919101/7f96ee76-c265-4b95-95fe-5fcf1d24166d

mwu-tow committed:
wip

mwu-tow committed:
Merge branch 'wip/mwu/new-ide-build' of https://github.com/luna/enso into wip/mwu/new-ide-build

mwu-tow committed:
wip

mwu-tow committed:
Merge remote-tracking branch 'origin/develop' into wip/mwu/new-ide-build

# Conflicts:
#	.vscode/settings.json
#	app/gui2/package-lock.json
#	app/gui2/vite.config.ts

mwu-tow committed:
wip

mwu-tow committed:
Merge branch 'wip/mwu/new-ide-build' of https://github.com/luna/enso into wip/mwu/new-ide-build

mwu-tow committed:
wip

mwu-tow committed:
wip

mwu-tow committed:
ignore

mwu-tow committed:
Merge remote-tracking branch 'origin/develop' into wip/mwu/new-ide-build

# Conflicts:
#	Cargo.lock
#	app/gui2/package-lock.json
#	app/gui2/package.json
#	app/gui2/rust-ffi/Cargo.toml
#	build/build/src/project/ide.rs
#	build/cli/src/lib.rs

mwu-tow committed:
wip

mwu-tow committed:
Merge remote-tracking branch 'origin/develop' into wip/mwu/new-ide-build

vitvakatu committed:
Force view update

vitvakatu committed:
Merge branch 'develop' into wip/vitvakatu/preview-context-switch

vitvakatu commented:
The issue was fixed, ready for another QA.

mwu-tow committed:
wip, cleanups

mwu-tow committed:
wip, cleanups

mwu-tow committed:
remove dep no longer necessary

mwu-tow committed:
more entries in the lock for @swc

mwu-tow committed:
swc for x64 mac

farmaazon commented:
## QA Report: :red_circle: 

Sorry for late testing, but I have problems with running IDE in electron (I'll describe them tomorrow). Tested in Chrome browser.

The last issue indeed is fixed, but still another remained (I think it wasn't there before, but I'm not 100% sure): Sometimes when I start editing, the changes are not propagated. In below screencast the first editions are not updated in code (seen in the code editor), but on second edit it works:



https://github.com/enso-org/enso/assets/3919101/f21da532-f5b3-4448-a67a-fed4ceac05aa


 I suspect that the non-working case is triggered when the opened CB has no entries.

mwu-tow committed:
separate job for building gui2

mwu-tow committed:
Merge remote-tracking branch 'origin/develop' into wip/mwu/new-ide-build

mwu-tow committed:
fixes for electron

mwu-tow committed:
don't run npm install in ensogl-pack

mwu-tow committed:
Allow explicit skipping of npm install. Use it in watch skcript.

mwu-tow committed:
fmt

mwu-tow committed:
CR

mwu-tow committed:
Merge remote-tracking branch 'origin/develop' into wip/mwu/new-ide-build

mwu-tow committed:
cleanups

mwu-tow committed:
fixing warnings

Frizi committed:
Merge branch 'develop' into wip/mwu/new-ide-build

Frizi committed:
Merge branch 'develop' into wip/mwu/new-ide-build

Frizi committed:
handle missing project metadata

Frizi committed:
add prettier ignore for a log file

Frizi committed:
trailing newline

mwu-tow committed:
CR feedback

mwu-tow committed:
Merge remote-tracking branch 'origin/wip/mwu/new-ide-build' into wip/mwu/new-ide-build

Frizi committed:
Merge branch 'develop' into wip/mwu/new-ide-build

Frizi committed:
lint

vitvakatu committed:
Merge branch 'develop' into wip/vitvakatu/preview-context-switch

vitvakatu committed:
Merge branch 'wip/mwu/new-ide-build' into wip/vitvakatu/preview-context-switch

vitvakatu committed:
Update preview even if no component were ever selected

vitvakatu committed:
Merge branch 'develop' into wip/vitvakatu/preview-context-switch

vitvakatu committed:
Update preview if visualization was already opened

vitvakatu committed:
Cleanup

vitvakatu commented:
I fixed the last issue with the visualization not updating when it was already opened before editing.


https://github.com/enso-org/enso/assets/6566674/7cc6afdb-bc83-42fe-af68-776df6998689

However, there are two more issues present, but they are also reproducible on develop: https://github.com/enso-org/enso/issues/7976 and https://github.com/enso-org/enso/issues/7977.

Besides, I also reported https://github.com/enso-org/enso/issues/7978 which also happens from time to time on the develop.

Other than that, this PR should be ready to be merged. @farmaazon

farmaazon commented on app/gui2/shared/languageServerTypes/suggestions.ts:
```diff
@@ -1,4 +1,4 @@
-import type { Uuid } from 'shared/languageServerTypes'
```
Seems like unintentional change

vitvakatu committed:
Revert merge conflict
